fw/cache.c

```diff
@@ -2986,14 +3121,32 @@ cache_req_process_node(TfwHttpReq *req, tfw_http_cache_cb_t action)
 		}
 	}
 
-	resp = tfw_cache_build_resp(req, ce, lifetime);
+	resp = tfw_cache_build_resp(req, ce, age);
```

Comment:
From my point of view, in this function we even can remove stale entries from the cache. However it will require architectural changes of htrie. `cache_req_process_node` knows whether entry alive or not by calling `tfw_cache_entry_is_live`, if entry is not alive we can remove it from the cache here in case when ability to response with stale responses is disabled.  At this moment we have two problems to solve:
1. We can't remove arbitrary element from the collision chain, because we need to lock previous element or whole chain, we can't lock previous element because our chain is single linked list. When we will turn it into double we need to verify all remove/insert logic, that we don't have deadlocks.
2. For stale responses we need refcounter or something else that give an ability to hold the record without locking and then remove this record if it was **successfully** revalidated.

Reply:
>  in this function we even can remove stale entries from the cache.

Yeah, but I'd move the eviction/GC to TDB: TDB is small while cache.c is already pretty big, so it'd be good to reduce code complexity in cache. Also client.c needs eviction as well.

Meantime, *revalidation* is a tricky thing - TDB can't just remove old entries, it must have some understanding on what's fine to evict and what's not. I think we can pass a callback to TDB, which will implement an eviction policy.

> We can't remove arbitrary element from the collision chain, because we need to lock previous element or whole chain, we can't lock previous element because our chain is single linked list. 

Can't we use some lock free approach and do additional spiining/checking on the list traversal?

Reply:
This comment is still not addressed

> Also I think we need to at least document the current design or maybe fix it. We construct a response from a stale cache entry and most likely the response will be just freed once we get a fresh response from the upstream, i.e. we do the construction with no reason

we need at least a new issue for this

Reply:
Implemented in new commit [Enhancement for cache_use_stale](https://github.com/tempesta-tech/tempesta/pull/2074/commits/85fa6c67c3089b967d41f698413d1ad1b5408af7)